tun/tun_solaris.go

```diff
@@ -0,0 +1,398 @@
+		msg := (*unix.IfMsghdr)(unsafe.Pointer(&data[0]))
+		if int(msg.Type) != unix.RTM_IFINFO {
+			continue
+		}
+		if int(msg.Index) != tunIfindex {
+			continue
+		}
+
+		iface, err := net.InterfaceByName(tun.name)
+		if err != nil {
+			tun.errors <- err
+			return
+		}
+
+		// Currently the solaris/illumos tun driver has two quirks that should be noted here
+		// 1. At creation time it immediately goes up before this code will have
+		// a chance to notice
+		// 2. It doesn't really respect 'ifconfig tunN down' properly so this code
+		// won't notice, and it will be broken and refuse to come back up
+		// if it is set down.
+		// If that ever changes, this code will be waiting.
+
+		// Up / Down event
+		up := (iface.Flags & net.FlagUp) != 0
+		if up != statusUp && up {
+			tun.events <- EventUp
+		}
+		if up != statusUp && !up {
+			tun.events <- EventDown
+		}
+		statusUp = up
+
+		// MTU changes
+		if iface.MTU != statusMTU {
+			tun.events <- EventMTUUpdate
+		}
+		statusMTU = iface.MTU
+	}
+}
+
+func tunDestroy(name string) error {
+	var l unix.Lifreq
+	if err := l.SetName(name); err != nil {
+		return err
+	}
+	// cleanup both ipv4 and ipv6
+	for _, proto := range []int{unix.AF_INET, unix.AF_INET6} {
+		fd, err := unix.Socket(proto, unix.SOCK_DGRAM, 0)
+		if err != nil {
+			return fmt.Errorf("could not open Socket of type %v", proto)
+		}
+		defer unix.Close(fd)
+
+		// get the STREAMS muxid for the tun device stream
+		reqnum := int(unix.SIOCGLIFMUXID)
+
+		// solaris/illumos defines the ioctl number as a signed int, but the
+		// common x/sys/unix functions use uint.
+		if err := unix.IoctlLifreq(fd, uint(reqnum), &l); err != nil {
+			continue
+		}
+
+		id := l.GetLifruInt()
+
+		// Unlink the tun stream from the IP mutiplexor
+		if err := unix.IoctlSetInt(fd, unix.I_PUNLINK, id); err != nil {
+			continue
+		}
+	}
+	return nil
+}
+
+func CreateTUN(name string, mtu int) (Device, error) {
+	ppa := -1
+	if name != "tun" {
+		_, err := fmt.Sscanf(name, "tun%d", &ppa)
+		if err != nil || ppa < 0 {
+			return nil, fmt.Errorf("interface name must be tun[0-9]*")
+		}
+	}
+
+	fd, err := unix.Open(tunNode, unix.O_RDWR, 0)
+	if err != nil {
+		return nil, fmt.Errorf("could not open TUN (%s)", tunNode)
+	}
+
+	// The "tun" device is a STREAMS module.  We need to make a STREAMS ioctl
+	// request to that module using the TUNNEWPPA command.  Returns the newly
+	// allocated PPA number. Passing -1 will request a dynamically assinged
+	// PPA. Passing a specific id will try to use the one specified.
+	var s unix.Strioctl
+	s.Cmd = unix.TUNNEWPPA
+	s.SetInt(ppa)
+	ppa, err = unix.IoctlSetStrioctlRetInt(fd, unix.I_STR, &s)
+	if err != nil {
+		return nil, err
+	}
+
+	assignedName := fmt.Sprintf("tun%d", ppa)
+
+	// Clean up anything possibly left behind by a crash
+	tunDestroy(assignedName)
+
+	// The solaris/illumos tun driver doesn't have an ioctl for reading back the PPA.
+	// Just pass it through to the daemon process via environment variable.
```

Comment:
Darn. Still? Weren't you talking to them about adding this at some point?

Reply:
Where's the source code for the tun driver? Is it not part of the base illumos?

Reply:
> Weren't you talking to them about adding this at some point?

No, that was other functionality. 

Also, I'm told that the code as it is today works on Solaris as well, an operating system that is closed source and over which I have even less influence. As gross as this is, I think it's a decent tradeoff to make to be able to support Solaris too.

If anything, some day illumos might get a newer better TUN driver that doesn't even use STREAMS and the solaris and illumos implementations in here would diverge. But that might not happen until I learn enough C to write one (and find the time and motivation to do so), so that could be (even more) years away...

Reply:
Can you link to the source code of the tun driver in illumos?

Reply:
As linked in a much older thread, the header file is here:
https://github.com/kaizawa/tuntap/blob/v1.3.3/if_tun.h

(Linking the version string referenced in e.g. the OmniOS build recipe:
https://github.com/omniosorg/omnios-build/blob/master/build/tuntap/build.sh#L21)

@gco or @alanc might be able to confirm that Solaris is using the same version...

Reply:
Solaris does not include the tuntap driver AFAIK, end-users need to install it if they need it.    @danmcd would probably know the history (and merged tuntap into illumos!).

Reply:
> Solaris does not include the tuntap driver AFAIK, end-users need to install it if they need it. @danmcd would probably know the history (and merged tuntap into illumos!).

It's not in-gate in illumos. The distros ship it separately.

I linked to the OmniOS build recipe above. The SmartOS one is https://github.com/TritonDataCenter/illumos-extra/tree/master/tun and also references 1.3.
The OpenIndiana one is at https://github.com/OpenIndiana/oi-userland/tree/oi/hipster/components/network/tuntap

I don't see the Solaris build recipe in https://github.com/oracle/solaris-userland/tree/master/components so I'm guessing that it's in a closed repo. I was just wondering what version string you folks have and whether you've made any interesting changes to it.

Reply:
I'll emphasize "does not include".   Solaris users need to get it from the 10 year old github repository.

Reply:
> I'll emphasize "does not include". Solaris users need to get it from the 10 year old github repository.

Got it. So to be very explicit, in order to use my wireguard-go and tailscale ports you installed the driver directly from the https://github.com/kaizawa/tuntap repo. Is that correct?

Reply:
I got mine from @danmcd many years ago as part of a private tool called punchin.  It's built from the same sources.

Reply:
Okay so it sounds like there's a single (old & crusty) codebase for tuntap shared across Solaris, Illumos, and derivatives. So,

> Also, I'm told that the code as it is today works on Solaris as well, an operating system that is closed source and over which I have even less influence. As gross as this is, I think it's a decent tradeoff to make to be able to support Solaris too.

This is not correct then, right?